nifi-extension-bundles/nifi-jms-bundle/nifi-jms-processors/src/main/java/org/apache/nifi/jms/processors/PublishJMS.java

```diff
@@ -211,8 +212,12 @@ public class PublishJMS extends AbstractJMSProcessor<JMSPublisher> {
      */
     @Override
     protected void rendezvousWithJms(ProcessContext context, ProcessSession processSession, JMSPublisher publisher) throws ProcessException {
-        FlowFile flowFile = processSession.get();
-        if (flowFile != null) {
+        final List<FlowFile> flowFiles = processSession.get(context.getProperty(MAX_BATCH_SIZE).asInteger());
+        if (flowFiles.isEmpty()) {
+            return;
+        }
+
+        flowFiles.forEach(flowFile -> {
```

Comment:
@mosermw Thanks for these enhancements to JMS consume and publish. 
Please review the existing code inside the new added loop and look for ways to optimize it. 
For example, move these variables to be created once before the loop if possible:
  allowIllegalChars
  attributeHeaderRegex 
  pattern
  readerFactory
  writerFactory
  flowFileReader

Reply:
Thanks for looking at this @jrsteinebrey.  Great suggestion.  I did this for the vars you mentioned except for flowFileReader.  Creating flowFileReader outside the loop would require checking the RECORD_READER property both outside and inside the loop.  I can do it if you feel strongly about it, but the code felt ugly when I had it in there like that.

Reply:
@mosermw Thanks for the changes. Your points on flowFileReader make perfect sense so I agree to leave that one as is.